Add BLAS level 1 routine snrm2

splrk opened the pull request:
Resolves #266.

<!--lint disable first-heading-level-->

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [X] Read, understood, and followed the [contributing guidelines][contributing], including the relevant style guides.
-   [X] Read and understand the [Code of Conduct][code-of-conduct].
-   [X] Read and understood the [licensing terms][license].
-   [X] Searched for existing issues and pull requests **before** submitting this pull request.
-   [X] Filed an issue (or an issue already existed) **prior to** submitting this pull request.
-   [X] Rebased onto latest `develop`.
-   [X] Submitted against `develop` branch.

## Description

> What is the purpose of this pull request?

This pull request adds the initial implementation of `@stdlib/blas/base/snrm2`.  WASM and native code glue not implemented

## Related Issues

> Does this pull request have any related issues?

This pull request:

-   resolves #266 

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

* * *

@stdlib-js/reviewers

<!-- <links> -->

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

[code-of-conduct]: https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md

[license]: https://github.com/stdlib-js/stdlib/blob/develop/LICENSE

<!-- </links> -->

Commit:
Add BLAS level 1 routine

splrk commented:
Sorry it's taken so long to get back into this.  I hope this is helpful.

kgryte commented:
@splrk Thanks for working on this!!! Will review shortly!

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/README.md:
```diff
@@ -0,0 +1,190 @@
+
+@license Apache-2.0
+
+Copyright (c) 2019 The Stdlib Authors.
```
Would you mind updating all the copyright years in this PR to `2020`? Thanks!

splrk replied:
No Problem

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/README.md:
```diff
@@ -0,0 +1,190 @@
+<!-- <equation class="equation" label="eq:l2norm" align="center raw="\|\mathbf{x}\|_2 = \sqrt{\mathbf{x}^T\mathbf{x}}" -->
+
+<div class="equation" align="center" data-raw-text="\|\mathbf{x}\|_2 = \sqrt{\mathbf{x}^T\mathbf{x}}" data-equation="eq:l2norm">
+    ![L2 norm equation][equation]
```
Ah. We do this a bit differently. See `ddot` for an example. Leave the `src` attribute as an empty string, and I'll update it on my end after merging this PR.

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/README.md:
```diff
@@ -0,0 +1,190 @@
+-   **stride**: index increment for `x`.
+
+The `N` and `stride` parameters determine which elements in `x` are accessed at runtime. For
+example, to compute the length of every other element of `x`,
```
You can remove the linebreak here.

kgryte replied:
```suggestion
example, to compute the 2-norm for every other element of `x`,
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/README.md:
```diff
@@ -0,0 +1,190 @@
+var Float32Array = require( '@stdlib/array/float32' );
+var floor = require( '@stdlib/math/base/special/floor' );
+
+var x = new Float32Array( [ 1.0, 2.0, 2.0, -7.0, -2.0, 3.0, 4.0, 2.0 ]);
```
```suggestion
var x = new Float32Array( [ 1.0, 2.0, 2.0, -7.0, -2.0, 3.0, 4.0, 2.0 ] );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/README.md:
```diff
@@ -0,0 +1,190 @@
+var floor = require( '@stdlib/math/base/special/floor' );
+
+var x = new Float32Array( [ 1.0, 2.0, 2.0, -7.0, -2.0, 3.0, 4.0, 2.0 ]);
+var N = floor(x.length / 2);
```
```suggestion
var N = floor( x.length / 2 );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/README.md:
````diff
@@ -0,0 +1,190 @@
+// returns 5.0
+```
+
+Note that indexing is relative to the first index.  To introduce and offset, use [`typed array`][mdn-typed-array] views.
````
```suggestion
Note that indexing is relative to the first index.  To introduce an offset, use [`typed array`][mdn-typed-array] views.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/README.md:
```diff
@@ -0,0 +1,190 @@
+var x0 = new Float32Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ]);
+var x1 = new Float32Array( x0.buffer, x0.BYTES_PER_ELEMENT*1 ); // start at 2nd element
+
+var N = floor(x.length / 2);
```
```suggestion
var N = floor( x.length / 2 );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/README.md:
```diff
@@ -0,0 +1,190 @@
+
+-   **offset**: starting index for `x`.
+
+While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the `offset` parameter supports indexing semantics based on starting indicies.  For example to calculate the 2-norm pf every other value in `x` starting from the second value
```
```suggestion
While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the `offset` parameter supports indexing semantics based on starting indicies.  For example, to calculate the 2-norm for every other value in `x` starting from the second value
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/README.md:
```diff
@@ -0,0 +1,190 @@
+
+var x = new Float32Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ]);
+
+var N = floor(x.length / 2);
```
```suggestion
var N = floor( x.length / 2 );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/README.md:
```diff
@@ -0,0 +1,190 @@
+
+## Notes
+
+-   If `N <= 0` both functions return `0.0`.
```
```suggestion
-   If `N <= 0` or `stride <= 0`, both functions return `0.0`.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/README.md:
```diff
@@ -0,0 +1,190 @@
+## Notes
+
+-   If `N <= 0` both functions return `0.0`.
+-   If `stride <= 0` both functions return `0.0`.
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/README.md:
```diff
@@ -0,0 +1,190 @@
+
+[snrm2]: http://www.netlib.org/lapack/explore-html/df/d28/group__single__blas__level1.html
+
+[equation]: https://latex.codecogs.com/svg.latex?%5Cinline%20%5C%7C%5Cmathbf%7Bx%7D%5C%7C_2%20%3D%20%5Csqrt%7B%5Cmathbf%7Bx%7D%5ET%5Cmathbf%7Bx%7D%7D
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/benchmark/c/Makefile:
```diff
@@ -0,0 +1,129 @@
+endif
+
+# List of source files:
+c_src := ../../src/snrm2.c ../../../../../math/base/special/abs/src/abs.c
```
```suggestion
c_src := ../../src/snrm2.c
```

kgryte replied:
We can rely on the built-in `abs`, rather than using the `stdlib` implementation. We'd also do this a bit differently, leveraging the `manifest.json` rather than explicitly referencing a source file.

splrk replied:
Awesome.  Thanks, I was wondering about that.  Is the `manifest.json` something used in compiling node native addons?

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/examples/index.js:
```diff
@@ -0,0 +1,38 @@
+var Float32Array = require( '@stdlib/array/float32' );
+var snrm2 = require( './../lib' ).ndarray;
+
+console.log(snrm2);
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/include/stdlib/blas/base/snrm2.h:
```diff
@@ -0,0 +1,42 @@
+#endif
+
+/**
+* Computes the 2-norm of a single-precision floating-pint vector.
```
```suggestion
* Computes the 2-norm of a single-precision floating-point vector.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/include/stdlib/blas/base/snrm2.h:
```diff
@@ -0,0 +1,42 @@
+#endif
+
+
+#endif // !SDOT_H
```
```suggestion
#endif // !SNRM2_H
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/include/stdlib/blas/base/snrm2_cblas.h:
```diff
@@ -0,0 +1,42 @@
+#endif
+
+/**
+* Computes the 2-norm of a single-precision floating-pint vector.
```
```suggestion
* Computes the 2-norm of a single-precision floating-point vector.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/index.js:
```diff
@@ -0,0 +1,58 @@
+
+/**
+*
+* Compute the 2-norm (length) of the given array.
```
```suggestion
* Compute the 2-norm (length) of an array.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/index.js:
```diff
@@ -0,0 +1,58 @@
+* @example
+* var snrm2 = require( '@stdlib/blas/base/snrm2' );
+*
+* var Float32Array = require( '@stdlib/array/float32' );
```
Move to above L28.

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/index.js:
```diff
@@ -0,0 +1,58 @@
+* // returns 3.0
+*
+* @example
+* var snrm2 = require( '@stdlib/blas/base/snrm2' );
```
Make the last import statement.

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/index.js:
```diff
@@ -0,0 +1,58 @@
+* var Float32Array = require( '@stdlib/array/float32' );
+* var floor = require( '@stdlib/math/base/special/floor' );
+*
+* var x = new Float32Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ]);
```
```suggestion
* var x = new Float32Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ] );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/index.js:
```diff
@@ -0,0 +1,58 @@
+*
+* var x = new Float32Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ]);
+*
+* var N = floor(x.length / 2);
```
```suggestion
* var N = floor( x.length / 2 );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/ndarray.js:
```diff
@@ -0,0 +1,91 @@
+/**
+* Compute the 2-norm of `x`.
+*
+* @param {PositiveInteger} N - number of values over which to copmute the 2-norm.
```
```suggestion
* @param {PositiveInteger} N - number of values over which to compute the 2-norm
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/ndarray.js:
```diff
@@ -0,0 +1,91 @@
+* Compute the 2-norm of `x`.
+*
+* @param {PositiveInteger} N - number of values over which to copmute the 2-norm.
+* @param {Float32Array} x - input array.
```
```suggestion
* @param {Float32Array} x - input array
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/ndarray.js:
```diff
@@ -0,0 +1,91 @@
+*
+* @param {PositiveInteger} N - number of values over which to copmute the 2-norm.
+* @param {Float32Array} x - input array.
+* @param {PositiveInteger} stride - `x` stride length.
```
```suggestion
* @param {PositiveInteger} stride - stride length
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/ndarray.js:
```diff
@@ -0,0 +1,91 @@
+* @param {PositiveInteger} N - number of values over which to copmute the 2-norm.
+* @param {Float32Array} x - input array.
+* @param {PositiveInteger} stride - `x` stride length.
+* @param {PositiveInteger} offset - starting index for `x`.
```
```suggestion
* @param {PositiveInteger} offset - starting index
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/ndarray.js:
```diff
@@ -0,0 +1,91 @@
+* var Float32Array = require( '@stdlib/array/float32' );
+* var floor = require( '@stdlib/math/base/special/floor' );
+*
+* var x = new Float32Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ]);
```
```suggestion
* var x = new Float32Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ] );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/ndarray.js:
```diff
@@ -0,0 +1,91 @@
+*
+* var x = new Float32Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ]);
+*
+* var N = floor(x.length / 2);
```
```suggestion
* var N = floor( x.length / 2 );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/ndarray.js:
```diff
@@ -0,0 +1,91 @@
+	if ( N <= 0 ) {
+		return 0.0;
+	}
+
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/ndarray.js:
```diff
@@ -0,0 +1,91 @@
+	if ( stride <= 0 ) {
+		return 0.0;
+	}
+
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/ndarray.js:
```diff
@@ -0,0 +1,91 @@
+	if ( N === 1 ) {
+		return abs( x[ 0 ] );
+	}
+
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/ndarray.js:
```diff
@@ -0,0 +1,91 @@
+	var ix;
+	var i;
+
+	if ( N <= 0 ) {
```
```suggestion
	if ( N <= 0 || stride <= 0 ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/ndarray.js:
```diff
@@ -0,0 +1,91 @@
+		return 0.0;
+	}
+
+	if ( stride <= 0 ) {
```
You can remove this conditional after consolidating with `N <= 0`.

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/ndarray.js:
```diff
@@ -0,0 +1,91 @@
+		}
+		ix += stride;
+	}
+
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/snrm2.js:
```diff
@@ -0,0 +1,88 @@
+/**
+* Compute the 2-norm of `x`.
+*
+* @param {PositiveInteger} N - number of values over which to copmute the 2-norm.
```
```suggestion
* @param {PositiveInteger} N - number of values over which to compute the 2-norm
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/snrm2.js:
```diff
@@ -0,0 +1,88 @@
+* Compute the 2-norm of `x`.
+*
+* @param {PositiveInteger} N - number of values over which to copmute the 2-norm.
+* @param {Float32Array} x - input array.
```
```suggestion
* @param {Float32Array} x - input array
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/snrm2.js:
```diff
@@ -0,0 +1,88 @@
+*
+* @param {PositiveInteger} N - number of values over which to copmute the 2-norm.
+* @param {Float32Array} x - input array.
+* @param {PositiveInteger} stride - `x` stride length.
```
```suggestion
* @param {PositiveInteger} stride - stride length
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/snrm2.js:
```diff
@@ -0,0 +1,88 @@
+	var ix;
+	var i;
+
+	if ( N <= 0 ) {
```
Same comments as for `ndarray.js`.

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/package.json:
```diff
@@ -0,0 +1,74 @@
+{
+  "name": "@stdlib/blas/base/snrm2",
+  "version": "0.0.0",
+  "description": "Calculates the 2-norm of vector x",
```
```suggestion
  "description": "Calculate the L2-norm of a vector.",
```

kgryte replied:
Also, would you mind updating all the uses of `2-norm` to `L2-norm`? Thanks!

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/package.json:
```diff
@@ -0,0 +1,74 @@
+    "euclidean norm",
+    "euclidean distance",
+    "2-norm",
+    "distance",
```
Can add `l2-norm`.

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2.c:
```diff
@@ -0,0 +1,72 @@
+* @see <a href="http://www.netlib.org/lapack/expolore-html/df/d28/group__single__blas__level1.html">sdot</a>
+*/
+#include <math.h>
+#include "stdlib_base_abs.h"
```
Can rely on the built-in `abs` function for this, rather than `stdlib`.

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2.c:
```diff
@@ -0,0 +1,72 @@
+/**
+* Compute the 2-norm of `X`.
+*
+* @see <a href="http://www.netlib.org/lapack/expolore-html/df/d28/group__single__blas__level1.html">sdot</a>
```
```suggestion
* @see <a href="http://www.netlib.org/lapack/expolore-html/df/d28/group__single__blas__level1.html">snrm2</a>
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2.c:
```diff
@@ -0,0 +1,72 @@
+#include "stdlib/blas/base/snrm2.h"
+
+/**
+* Computes the 2-norm of a single-precision floating-pint vector.
```
```suggestion
* Computes the L2-norm of a single-precision floating-point vector.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2.c:
```diff
@@ -0,0 +1,72 @@
+/**
+* Computes the 2-norm of a single-precision floating-pint vector.
+*
+* @param N       number of values over which to compute the norm.
```
```suggestion
* @param N       number of values over which to compute the norm
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2.c:
```diff
@@ -0,0 +1,72 @@
+	}
+
+	if ( N == 1 ) {
+		return stdlib_base_abs( X[ 0 ] );
```
```suggestion
		return abs( X[ 0 ] );
```

splrk replied:
Using `fabs` instead since `abs` is for `int` only

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2.c:
```diff
@@ -0,0 +1,72 @@
+	ssq = 1.0;
+	for ( i = 0; i < N; i += 1 ) {
+		if ( X[ ix ] != 0.0 ) {
+			absxi = stdlib_base_abs( X[ ix ] );
```
```suggestion
			absxi = abs( X[ ix ] );
```

splrk replied:
Same as for line 52

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2.c:
```diff
@@ -0,0 +1,72 @@
+		return 0.0;
+	}
+
+	if ( stride <= 0 ) {
```
Consolidate with prior conditional.

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2.f:
```diff
@@ -0,0 +1,102 @@
+! limitations under the License.
+!<
+
+!> Computes the 2-norm of a vector
```
```suggestion
!> Computes the L2-norm of a vector.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2.f:
```diff
@@ -0,0 +1,102 @@
+!
+! * Jack Dongarra, linpack, 3/11/78.
+!
+!   - modified 12/3/93, array(1) declarations changed to array(*)
```
```suggestion
```

kgryte replied:
And also remove the surrounding linebreaks.

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2.f:
```diff
@@ -0,0 +1,102 @@
+!
+! ## History
+!
+! * Jack Dongarra, linpack, 3/11/78.
```
```suggestion
! * Original version written on 25-October-1982.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2.f:
```diff
@@ -0,0 +1,102 @@
+! @param {integer} N - number of values over which to compute the nrom
+! @param {Array<real>} sx - array
+! @param {integer} strideX - `sx` stride length
+! @returns {real} the nrom of `x`
```
```suggestion
! @returns {real} the norm of `x`
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2.f:
```diff
@@ -0,0 +1,102 @@
+real function snrm2( N, x, stride )
+  implicit none
+  !..
+  ! Scalar argurments
```
```suggestion
  ! Scalar arguments:
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2.f:
```diff
@@ -0,0 +1,102 @@
+  ! Scalar argurments
+  integer :: N, stride
+  ! ..
+  ! Array arguments
```
```suggestion
  ! Array arguments:
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2.f:
```diff
@@ -0,0 +1,102 @@
+  ! Array arguments
+  real, intent(in) :: x(*)
+  ! ..
+  ! Local Scalars
```
```suggestion
  ! Local scalars:
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2.f:
```diff
@@ -0,0 +1,102 @@
+  real :: absxi, scale, ssq
+  integer :: i, ix
+  ! ..
+  ! Intrinsic functions
```
```suggestion
  ! Intrinsic functions:
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2_cblas.c:
```diff
@@ -0,0 +1,39 @@
+* @see <a href="http://www.netlib.org/lapack/expolore-html/df/d28/group__single__blas__level1.html">snrm2</a>
+*/
+#include <math.h>
+#include "stdlib_base_abs.h"
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2_cblas.c:
```diff
@@ -0,0 +1,39 @@
+#include "stdlib/blas/base/snrm2_cblas.h"
+
+/**
+* Computes the 2-norm of a single-precision floating-pint vector.
```
```suggestion
* Computes the 2-norm of a single-precision floating-point vector.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2_cblas.c:
```diff
@@ -0,0 +1,39 @@
+/**
+* Computes the 2-norm of a single-precision floating-pint vector.
+*
+* @param N       number of values over which to compute the norm.
```
```suggestion
* @param N       number of values over which to compute the norm
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2_cblas.c:
```diff
@@ -0,0 +1,39 @@
+* @param stideX  stride length
+* @return        output value
+*/
+float c_snrm2( const int N, const float *X, const int stride) {
```
```suggestion
float c_snrm2( const int N, const float *X, const int stride ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2_f.c:
```diff
@@ -0,0 +1,40 @@
+#include "stdlib/blas/base/snrm2_fortran.h"
+
+/**
+* Computes the 2-norm of a single-precision floating-pint vector.
```
```suggestion
* Computes the 2-norm of a single-precision floating-point vector.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2_f.c:
```diff
@@ -0,0 +1,40 @@
+/**
+* Computes the 2-norm of a single-precision floating-pint vector.
+*
+* @param N       number of values over which to compute the norm.
```
```suggestion
* @param N       number of values over which to compute the norm
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2_f.c:
```diff
@@ -0,0 +1,40 @@
+*
+* @param N       number of values over which to compute the norm.
+* @param X       input array
+* @param stideX  stride length
```
```suggestion
* @param stride  stride length
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2sub.f:
```diff
@@ -0,0 +1,46 @@
+!<
+
+!> Wraps snrm2 as a subroutine.
+
```
```suggestion
!
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2sub.f:
```diff
@@ -0,0 +1,46 @@
+! limitations under the License.
+!<
+
+!> Wraps snrm2 as a subroutine.
```
```suggestion
!> Wraps `snrm2` as a subroutine.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2sub.f:
```diff
@@ -0,0 +1,46 @@
+      real :: x(*)
+  end interface
+  ! ..
+  ! Scalar arguments
```
```suggestion
  ! Scalar arguments:
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2sub.f:
```diff
@@ -0,0 +1,46 @@
+  integer :: stride, none
+  real, intent(out) :: nrm2
+  ! ..
+  ! Array arguments
```
```suggestion
  ! Array arguments:
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2sub.f:
```diff
@@ -0,0 +1,46 @@
+  end interface
+  ! ..
+  ! Scalar arguments
+  integer :: stride, none
```
```suggestion
  integer :: stride, N
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2sub.f:
```diff
@@ -0,0 +1,46 @@
+  ! Array arguments
+  real :: x(*)
+  ! ..
+  ! Compute the 2-norm
```
```suggestion
  ! Compute the L2-norm:
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/test/test.js:
```diff
@@ -0,0 +1,38 @@
+	t.end();
+});
+
+tape( 'attached to the main export is a mthoed providing an ndarray interface', function test( t ) {
```
```suggestion
tape( 'attached to the main export is a method providing an ndarray interface', function test( t ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/test/test.ndarray.js:
```diff
@@ -0,0 +1,130 @@
+	t.end();
+});
+
+tape( 'the function calculates the 2-norm of arrays `x`', function test( t ) {
```
```suggestion
tape( 'the function calculates the L2-norm of a vector', function test( t ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/test/test.snrm2.js:
```diff
@@ -0,0 +1,132 @@
+	t.end();
+});
+
+tape( 'the function calculates the 2-norm of arrays `x`', function test( t ) {
```
```suggestion
tape( 'the function calculates the L2-norm of a vector', function test( t ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/test/test.snrm2.js:
```diff
@@ -0,0 +1,132 @@
+		4.0,   // 3
+		6.0
+	]);
+	x1 = new Float32Array( x0.buffer, x0.BYTES_PER_ELEMENT*1 ); // start at 2nd BYTES_PER_ELEMENTt
```
```suggestion
	x1 = new Float32Array( x0.buffer, x0.BYTES_PER_ELEMENT*1 ); // start at 2nd element
```

kgryte reviewed (changes requested):
@splrk Thanks again for working on this! Most of my comments are smaller nitpicks, and, where possible, I have used GitHub's suggestion feature to make incorporating the changes as straightforward as possible. Once these requested changes are addressed, this PR should be ready for merge!

kgryte reviewed (changes requested):
@splrk Thanks again for working on this! Most of my comments are smaller nitpicks, and, where possible, I have used GitHub's suggestion feature to make incorporating the changes as straightforward as possible. Once these requested changes are addressed, this PR should be ready for merge!

splrk committed:
Apply changes from code review

Change Copyright to 2020

cleanup readme

Update lib/node_modules/@stdlib/blas/base/snrm2/README.md

Co-Authored-By: Athan <kgryte@gmail.com>

Update lib/node_modules/@stdlib/blas/base/snrm2/README.md

Co-Authored-By: Athan <kgryte@gmail.com>

Update lib/node_modules/@stdlib/blas/base/snrm2/README.md

Co-Authored-By: Athan <kgryte@gmail.com>

Apply suggestions from code review

Clean up Readme

Co-Authored-By: Athan <kgryte@gmail.com>

Apply suggestions from code review

Co-Authored-By: Athan <kgryte@gmail.com>

Change 2-norm to L2-norm

Apply changes from code review

Apply suggestions from code review

Co-Authored-By: Athan <kgryte@gmail.com>

Apply changes from code review

Apply changes from code review

Apply changes from code review

splrk committed:
Apply changes from code review

Change Copyright to 2020

cleanup readme

Update lib/node_modules/@stdlib/blas/base/snrm2/README.md

Co-Authored-By: Athan <kgryte@gmail.com>

Update lib/node_modules/@stdlib/blas/base/snrm2/README.md

Co-Authored-By: Athan <kgryte@gmail.com>

Update lib/node_modules/@stdlib/blas/base/snrm2/README.md

Co-Authored-By: Athan <kgryte@gmail.com>

Apply suggestions from code review

Clean up Readme

Co-Authored-By: Athan <kgryte@gmail.com>

Apply suggestions from code review

Co-Authored-By: Athan <kgryte@gmail.com>

Change 2-norm to L2-norm

Apply changes from code review

Apply suggestions from code review

Co-Authored-By: Athan <kgryte@gmail.com>

Apply changes from code review

Apply changes from code review

Apply changes from code review

splrk commented:
Thanks for the quick review.  I think I got everything.

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/README.md:
```diff
@@ -0,0 +1,186 @@
+
+<section class="intro>">
+
+The [L2 norm][l2norm] is defined as
```
```suggestion
The [L2-norm][l2-norm] is defined as
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/README.md:
```diff
@@ -0,0 +1,186 @@
+
+The [L2 norm][l2norm] is defined as
+
+<!-- <equation class="equation" label="eq:l2norm" align="center raw="\|\mathbf{x}\|_2 = \sqrt{\mathbf{x}^T\mathbf{x}}" -->
```
```suggestion
<!-- <equation class="equation" label="eq:l2_norm" align="center raw="\|\mathbf{x}\|_2 = \sqrt{\mathbf{x}^T\mathbf{x}}" -->
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/README.md:
```diff
@@ -0,0 +1,186 @@
+
+<!-- <equation class="equation" label="eq:l2norm" align="center raw="\|\mathbf{x}\|_2 = \sqrt{\mathbf{x}^T\mathbf{x}}" -->
+
+<div class="equation" align="center" data-raw-text="\|\mathbf{x}\|_2 = \sqrt{\mathbf{x}^T\mathbf{x}}" data-equation="eq:l2norm">
```
```suggestion
<div class="equation" align="center" data-raw-text="\|\mathbf{x}\|_2 = \sqrt{\mathbf{x}^T\mathbf{x}}" data-equation="eq:l2_norm">
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/README.md:
```diff
@@ -0,0 +1,186 @@
+<!-- <equation class="equation" label="eq:l2norm" align="center raw="\|\mathbf{x}\|_2 = \sqrt{\mathbf{x}^T\mathbf{x}}" -->
+
+<div class="equation" align="center" data-raw-text="\|\mathbf{x}\|_2 = \sqrt{\mathbf{x}^T\mathbf{x}}" data-equation="eq:l2norm">
+    <img src="" alt="L2 norm definition.">
```
```suggestion
    <img src="" alt="L2-norm definition.">
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/README.md:
```diff
@@ -0,0 +1,186 @@
+
+#### snrm2( N, x, stride )
+
+Computes the L2-norm (length) of the given array.
```
```suggestion
Computes the [L2-norm][l2-norm] of the given array.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/README.md:
```diff
@@ -0,0 +1,186 @@
+The function accepts the following parameters:
+
+-   **N**: number of elements in the array.
+-   **x**: input [`Float32Array`][@stdlib/array/float32]
```
```suggestion
-   **x**: input [`Float32Array`][@stdlib/array/float32].
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/README.md:
```diff
@@ -0,0 +1,186 @@
+-   **x**: input [`Float32Array`][@stdlib/array/float32]
+-   **stride**: index increment for `x`.
+
+The `N` and `stride` parameters determine which elements in `x` are accessed at runtime. For example, to compute the length of every other element of `x`,
```
```suggestion
The `N` and `stride` parameters determine which elements in `x` are accessed at runtime. For example, to compute the [L2-norm][l2-norm] of every other element of `x`,
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/README.md:
```diff
@@ -0,0 +1,186 @@
+
+#### snrm2.ndarray( N, x, stride, offset )
+
+Computes the L2-norm (length) of the given array with alternative indexing semantics.
```
```suggestion
Computes the [L2-norm][l2-norm] of the given array with alternative indexing semantics.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/README.md:
```diff
@@ -0,0 +1,186 @@
+
+-   **offset**: starting index for `x`.
+
+While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the `offset` parameter supports indexing semantics based on starting indicies.  For example, to calculate the L2-norm for every other value in `x` starting from the second value
```
```suggestion
While [`typed array`][mdn-typed-array] views mandate a view offset based on the underlying `buffer`, the `offset` parameter supports indexing semantics based on starting indicies.  For example, to calculate the [L2-norm][l2-norm] for every other value in `x` starting from the second value
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/README.md:
```diff
@@ -0,0 +1,186 @@
+var floor = require( '@stdlib/math/base/special/floor' );
+
+var x = new Float32Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ]);
+
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/README.md:
```diff
@@ -0,0 +1,186 @@
+
+<section class="links">
+
+[l2norm]: https://en.wikipedia.org/wiki/Euclidean_distance
```
```suggestion
[l2-norm]: https://en.wikipedia.org/wiki/Euclidean_distance
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,151 @@
+}
+
+/**
+* Generates a random double on the interval [0,1].
```
```suggestion
* Generates a random number on the interval [0,1].
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,151 @@
+/**
+* Generates a random double on the interval [0,1].
+*
+* @return random double
```
```suggestion
* @return random number
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/examples/index.js:
```diff
@@ -0,0 +1,36 @@
+
+x = new Float32Array( 10 );
+for ( i = 0; i < x.length; i++ ) {
+	x[ i ] = round( randu() * 10 );
```
```suggestion
	x[ i ] = round( randu() * 10.0 );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/include/stdlib/blas/base/snrm2_fortran.h:
```diff
@@ -0,0 +1,41 @@
+}
+#endif
+
+#endif // !SNRM2_FORTRAN_H
```
```suggestion
#endif // !SNRM2_FORTRAN_H

```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/ndarray.js:
```diff
@@ -0,0 +1,87 @@
+	if ( N <= 0 || stride <= 0 ) {
+		return 0.0;
+	}
+	if ( stride <= 0 ) {
```
This conditional should be deleted.

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/snrm2.js:
```diff
@@ -0,0 +1,88 @@
+		return 0.0;
+	}
+
+	if ( stride <= 0 ) {
```
This conditional should be consolidated with `N <= 0` above.

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2.c:
```diff
@@ -0,0 +1,63 @@
+	ix = 0;
+	scale = 0.0;
+	ssq = 1.0;
+	for ( i = 0; i < N; i += 1 ) {
```
```suggestion
	for ( i = 0; i < N; i++ ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2.f:
```diff
@@ -0,0 +1,102 @@
+! limitations under the License.
+!<
+
+!> Computes the L2-norm of a vector
```
```suggestion
!> Computes the L2-norm of a vector.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2.f:
```diff
@@ -0,0 +1,102 @@
+!
+! * Original version written on 25-October-1982.
+!
+!   - modified 12/3/93, array(1) declarations changed to array(*)
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2_cblas.c:
```diff
@@ -0,0 +1,38 @@
+*
+* @param N       number of values over which to compute the norm
+* @param X       input array
+* @param stideX  stride length
```
```suggestion
* @param stride  stride length
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2sub.f:
```diff
@@ -0,0 +1,46 @@
+  ! ..
+  ! Scalar arguments:
+  integer :: stride, N
+  real, intent(out) :: nrm2
```
```suggestion
  real :: nrm2
```

To my knowledge, we don't need to indicate the `intent` here.

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2sub.f:
```diff
@@ -0,0 +1,46 @@
+  ! Array arguments:
+  real :: x(*)
+  ! ..
+  ! Compute the L2-norm
```
```suggestion
  ! Compute the L2-norm:
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/test/test.snrm2.js:
```diff
@@ -0,0 +1,132 @@
+	t.end();
+});
+
+tape( 'the function calculates the L2-norm of arrays `x`', function test( t ) {
```
```suggestion
tape( 'the function calculates the L2-norm of a vector', function test( t ) {
```

kgryte reviewed (changes requested):
@splrk Mainly a few missing punctuations and minor changes. Thanks for your continued work on this!

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/benchmark/c/benchmark.length.c:
```diff
@@ -0,0 +1,151 @@
+*
+* @return random double
+*/
+double rand_float() {
```
```suggestion
float rand_float() {
```

kgryte committed:
Apply suggestions from code review

kgryte commented:
@splrk I think the only remaining changes are the consolidation of conditionals.

splrk commented:
Sorry, I didn't see you comments, please ignore the "re-review" requested

kgryte commented:
@splrk No worries!

Commit:
Merge branch 'feature/snrm2' of github.com:splrk/stdlib into feature/snrm2

splrk commented:
> @splrk I think the only remaining changes are the consolidation of conditionals.

Okay, 

> @splrk I think the only remaining changes are the consolidation of conditionals.

Yeah, I forgot to check the other files for this.  I went through and double checked.  These should be consolidated now.

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/README.md:
```diff
@@ -0,0 +1,185 @@
+var Float32Array = require( '@stdlib/array/float32' );
+var floor = require( '@stdlib/math/base/special/floor' );
+
+var x = new Float32Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ]);
```
```suggestion
var x = new Float32Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ] );
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/include/stdlib/blas/base/snrm2_fortran.h:
```diff
@@ -0,0 +1,41 @@
+#endif
+
+/**
+* Computes the dot product of two single-precision floating-point vectors.
```
```suggestion
* Computes the L2-norm of a single-precision floating-point vector.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/index.js:
```diff
@@ -0,0 +1,57 @@
+* var snrm2 = require( '@stdlib/blas/base/snrm2' );
+*
+* var x = new Float32Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ] );
+*
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/ndarray.js:
```diff
@@ -0,0 +1,84 @@
+* var floor = require( '@stdlib/math/base/special/floor' );
+*
+* var x = new Float32Array( [ 2.0, 1.0, 2.0, -2.0, -2.0, 2.0, 3.0, 4.0 ] );
+*
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/ndarray.js:
```diff
@@ -0,0 +1,84 @@
+	ix = offset;
+	scale = 0.0;
+	ssq = 1.0;
+	for ( i = 0; i < N; i += 1 ) {
```
```suggestion
	for ( i = 0; i < N; i++ ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/snrm2.js:
```diff
@@ -0,0 +1,82 @@
+	ix = 0;
+	scale = 0.0;
+	ssq = 1.0;
+	for ( i = 0; i < N; i += 1 ) {
```
```suggestion
	for ( i = 0; i < N; i++ ) {
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/snrm2.js:
```diff
@@ -0,0 +1,82 @@
+		}
+		ix += stride;
+	}
+
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/index.js:
```diff
@@ -0,0 +1,57 @@
+
+/**
+*
+* Compute the L2-norm (length) of an array.
```
```suggestion
* Compute the L2-norm of a vector.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/snrm2.js:
```diff
@@ -0,0 +1,82 @@
+// MAIN //
+
+/**
+* Compute the L2-norm of `x`.
```
```suggestion
* Compute the L2-norm of a vector.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/lib/ndarray.js:
```diff
@@ -0,0 +1,84 @@
+// MAIN //
+
+/**
+* Compute the L2-norm of `x`.
```
```suggestion
* Compute the L2-norm of a vector.
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2.c:
```diff
@@ -0,0 +1,63 @@
+*
+* @param N       number of values over which to compute the norm
+* @param X       input array
+* @param stideX  stride length
```
```suggestion
* @param stride  stride length
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2.f:
```diff
@@ -0,0 +1,97 @@
+!
+! * Original version written on 25-October-1982.
+!
+!
```
```suggestion
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2.f:
```diff
@@ -0,0 +1,97 @@
+!
+! @param {integer} N - number of values over which to compute the nrom
+! @param {Array<real>} sx - array
+! @param {integer} strideX - `sx` stride length
```
```suggestion
! @param {integer} stride - stride length
```

kgryte commented on lib/node_modules/@stdlib/blas/base/snrm2/src/snrm2.f:
```diff
@@ -0,0 +1,97 @@
+! >
+! > * We will gladly answer any questions regarding the software. If a modification is done, however, it is the responsibility of the person who modified the routine to provide support.
+!
+! @param {integer} N - number of values over which to compute the nrom
```
```suggestion
! @param {integer} N - number of values over which to compute the norm
```

kgryte committed:
Apply suggestions from code review
